src/plugins/hidsim/gui/vmi_win_gui_parser.h

```diff
@@ -0,0 +1,218 @@
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Jan Gruber.                                    *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+#ifndef VMI_WIN_GUI_PARSER_H
+#define VMI_WIN_GUI_PARSER_H
+
+/* Datastructures */
+#include <glib.h>
+
+#include <libvmi/libvmi.h>
+
+/* Flags indicating window style */
+#define WS_MINIMIZE 0x20000000
+#define WS_VISIBLE 0x10000000
+#define WS_DISABLED 0x08000000
+
+/* Window extended style (exstyle) */
+#define WS_EX_DLGMODALFRAME 0x00000001
+#define WS_EX_NOPARENTNOTIFY 0x00000004
+#define WS_EX_TOPMOST 0x00000008
+#define WS_EX_ACCEPTFILES 0x00000010
+#define WS_EX_TRANSPARENT 0x00000020
+
+/* Factor of desktop - button, used for simple filtering */
+#define BTN_RATIO 4
+
+/* Holds struct-offsets, needed to access relevant fields */
+extern struct Offsets symbol_offsets;
+
+/*
+ * The following structs encapsulate only the information needed for the purpose
+ * of reconstructing the GUI to a level, where dialogs could be identified for
+ * clicking
+ */
+struct winsta
+{
+    addr_t addr;
+    /*
+     * For each GUI thread, win32k maps, the associated desktop heap into
+     * usermode http://mista.nu/research/mandt-win32k-slides.pdf
+     *
+     * Therefore do it like volatility: Find a process with matching sessionID
+     * and take its VA as _MM_SESSION_SPACE for the WinSta
+     * https://github.com/volatilityfoundation/volatility/blob/\
+     * a438e768194a9e05eb4d9ee9338b881c0fa25937/volatility/plugins/\
+     * gui/sessions.py#L49
+     *
+     * To accomplish this, it's the most easy way to use vmi_read_xx_va
+     */
+    vmi_pid_t providing_pid;
+    uint32_t session_id;
+    addr_t atom_table;
+    bool is_interactive;
+    size_t len_desktops;
+    addr_t* desktops;
+    char* name;
+};
+
+/* Just the minimal information needed to retrieve windows */
+struct desktop
+{
+    addr_t addr;
+    vmi_pid_t providing_pid;
+    char* name;
+#ifndef DISABLE_ATOMS
+    GHashTable* atom_table;
+#endif
+};
+
+struct rect
+{
+    int16_t x0;
+    int16_t x1;
+    int16_t y0;
+    int16_t y1;
+
+    /* For convenience */
+    uint16_t w;
+    uint16_t h;
```

Comment:
Are those used anywhere? I can't find any usage apart from setting the fields.

OTOH, adding redundant fields is IMO a bad idea.
Personally I'd use x,y coordinates and width/height instead of two point, but I guess it's a matter of preference.

Reply:
Those are used in `gui_monitor.cpp`, l. 421 and in `vmi_win_gui_parser.cpp` in `find_button_to_click`. I introduced those with reasoning. I'd prefer to stick with this implementation.